File: wear/src/main/java/com/pdyjak/powerampwear/MainActivity.java
package com.pdyjak.powerampwear;

import android.graphics.Bitmap;
import android.graphics.BitmapFactory;
import android.os.Bundle;
import android.os.Handler;
import android.os.Looper;
import android.support.annotation.NonNull;
import android.support.v4.view.ViewPager;
import android.support.wearable.activity.WearableActivity;
import android.view.View;
import android.widget.ImageView;

import com.google.android.gms.wearable.MessageEvent;
import com.pdyjak.powerampwear.music_browser.files.FileItem;
import com.pdyjak.powerampwearcommon.events.AlbumArtChangedEvent;
import com.pdyjak.powerampwearcommon.events.TrackChangedEvent;

import me.relex.circleindicator.CircleIndicator;

public class MainActivity extends WearableActivity implements MessageListener {

    public static final int DOTS_VISIBILITY_TIMEOUT = 1000;
    public static final int OFFSCREEN_PAGE_LIMIT = 2;

    private class MusicBrowserListener extends MusicBrowserListenerAdapter {
        @Override
        public void onFileSelected(@NonNull FileItem item) {
            mShouldNavigateToPlayerView = true;
        }
    }

    private class PageChangeListener implements ViewPager.OnPageChangeListener {
        @Override
        public void onPageScrolled(int position, float positionOffset, int positionOffsetPixels) {
        }

        @Override
        public void onPageSelected(int position) {
        }

        @Override
        public void onPageScrollStateChanged(int state) {
            if (state == ViewPager.SCROLL_STATE_IDLE) {
                if (mHideDotsIndicatorRunnable == null) {
                    mHideDotsIndicatorRunnable = new Runnable() {
                        @Override
                        public void run() {
                            mDotsIndicator.setVisibility(View.GONE);
                        }
                    };
                    mHandler.postDelayed(mHideDotsIndicatorRunnable, DOTS_VISIBILITY_TIMEOUT);
                }
            } else {
                if (mHideDotsIndicatorRunnable != null) {
                    mHandler.removeCallbacks(mHideDotsIndicatorRunnable);
                    mHideDotsIndicatorRunnable = null;
                }
                mDotsIndicator.setVisibility(View.VISIBLE);
            }
        }
    }

    @NonNull
    private final Handler mHandler = new Handler(Looper.getMainLooper());
    @NonNull
    private final MusicBrowserListener mMusicBrowserListener = new MusicBrowserListener();

    private ImageView mAlbumArt;
    private Bitmap mAlbumArtBitmap;
    private boolean mShouldNavigateToPlayerView;
    private ViewPager mViewPager;
    private CircleIndicator mDotsIndicator;
    private Runnable mHideDotsIndicatorRunnable;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setAmbientEnabled();
        setContentView(R.layout.activity_main);
        mViewPager = (ViewPager) findViewById(R.id.view_pager);
        mViewPager.setAdapter(new MainViewPagerAdapter(getFragmentManager()));
        mViewPager.addOnPageChangeListener(new PageChangeListener());
        mViewPager.setOffscreenPageLimit(OFFSCREEN_PAGE_LIMIT);
        mAlbumArt = (ImageView) findViewById(R.id.album_art);
        mDotsIndicator = (CircleIndicator) findViewById(R.id.dots_indicator);
        mDotsIndicator.setViewPager(mViewPager);
    }

    @Override
    public void onEnterAmbient(Bundle ambientDetails) {
        super.onEnterAmbient(ambientDetails);
        updateDisplay();
    }

    @Override
    public void onExitAmbient() {
        super.onExitAmbient();
        updateDisplay();
    }

    @Override
    protected void onResume() {
        App app = (App) getApplicationContext();
        app.onResume();
        app.getMessageExchangeHelper().addMessageListenerWeakly(this);
        app.getMusicLibraryNavigator().addLibraryNavigationListener(mMusicBrowserListener);
        super.onResume();
    }

    @Override
    protected void onPause() {
        App app = (App) getApplicationContext();
        app.onPause();
        app.getMessageExchangeHelper().removeMessageListener(this);
        app.getMusicLibraryNavigator().removeLibraryNavigationListener(mMusicBrowserListener);
        super.onPause();
    }

    @Override
    public void onMessageReceived(@NonNull MessageEvent messageEvent) {
        if (AlbumArtChangedEvent.PATH.equals(messageEvent.getPath())) {
            byte[] bytes = messageEvent.getData();
            if (bytes == null) {
                mAlbumArtBitmap = null;
            } else {
                mAlbumArtBitmap = BitmapFactory.decodeByteArray(bytes, 0, bytes.length);
            }
            updateDisplay();
        } else if (TrackChangedEvent.PATH.equals(messageEvent.getPath())
                && mShouldNavigateToPlayerView) {
            mShouldNavigateToPlayerView = false;
            mViewPager.setCurrentItem(0, true);
        }
    }

    private void updateDisplay() {
        mAlbumArt.setImageBitmap(isAmbient() ? null : mAlbumArtBitmap);
    }
}

File: wear/src/main/java/com/pdyjak/powerampwear/MainActivity.kt
package com.pdyjak.powerampwear

import android.content.Context
import android.graphics.Bitmap
import android.graphics.BitmapFactory
import android.os.Bundle
import android.os.Handler
import android.os.Looper
import android.support.v4.view.ViewPager
import android.support.wearable.activity.WearableActivity
import android.view.View
import android.widget.ImageView
import com.google.android.gms.wearable.MessageEvent
import com.pdyjak.powerampwear.common.ambientModeStateProvider
import com.pdyjak.powerampwear.common.app
import com.pdyjak.powerampwear.music_browser.Clickable
import com.pdyjak.powerampwear.music_browser.FileSelectedEventArgs
import com.pdyjak.powerampwear.music_browser.ItemSelectedEventArgs
import com.pdyjak.powerampwearcommon.events.AlbumArtChangedEvent
import com.pdyjak.powerampwearcommon.events.TrackChangedEvent
import me.relex.circleindicator.CircleIndicator

class MainActivity : WearableActivity(), MessageListener {

    companion object {
        const val DOTS_VISIBILITY_TIMEOUT: Long = 1000
        const val OFFSCREEN_PAGE_LIMIT = 2
    }

    private val mFileSelectedEventHandler = { _: FileSelectedEventArgs ->
        mShouldNavigateToPlayerView = true
    }

    private val mItemSelectedEventHandler = { args: ItemSelectedEventArgs<Clickable>? ->
        if (args!!.fromPlayer) mViewPager!!.setCurrentItem(1, true)
    }

    private inner class PageChangeListener : ViewPager.OnPageChangeListener {
        override fun onPageScrolled(position: Int, positionOffset: Float,
                                    positionOffsetPixels: Int) {}

        override fun onPageSelected(position: Int) {}

        override fun onPageScrollStateChanged(state: Int) {
            if (state == ViewPager.SCROLL_STATE_IDLE) {
                if (mHideDotsIndicatorRunnable === null) {
                    mHideDotsIndicatorRunnable = Runnable {
                        mDotsIndicator!!.visibility = View.GONE
                    }
                    mHandler.postDelayed(mHideDotsIndicatorRunnable, DOTS_VISIBILITY_TIMEOUT)
                }
            } else {
                if (mHideDotsIndicatorRunnable !== null) {
                    mHandler.removeCallbacks(mHideDotsIndicatorRunnable)
                    mHideDotsIndicatorRunnable = null
                }
                mDotsIndicator!!.visibility = View.VISIBLE
            }
        }
    }

    private val mHandler = Handler(Looper.getMainLooper())

    private var mAlbumArt: ImageView? = null
    private var mViewPager: ViewPager? = null
    private var mDotsIndicator: CircleIndicator? = null
    private var mAlbumArtBitmap: Bitmap? = null
    private var mShouldNavigateToPlayerView: Boolean = false
    private var mHideDotsIndicatorRunnable: Runnable? = null

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setAmbientEnabled()
        setContentView(R.layout.activity_main)
        mViewPager = findViewById(R.id.view_pager) as ViewPager
        mViewPager!!.adapter = MainViewPagerAdapter(fragmentManager)
        mViewPager!!.addOnPageChangeListener(PageChangeListener())
        mViewPager!!.offscreenPageLimit = OFFSCREEN_PAGE_LIMIT
        mAlbumArt = findViewById(R.id.album_art) as ImageView
        mDotsIndicator = findViewById(R.id.dots_indicator) as CircleIndicator
        mDotsIndicator!!.setViewPager(mViewPager)
    }

    override fun onEnterAmbient(ambientDetails: Bundle?) {
        super.onEnterAmbient(ambientDetails)
        ambientModeStateProviderInternal.isInAmbientMode = true
        updateDisplay()
    }

    override fun onExitAmbient() {
        super.onExitAmbient()
        ambientModeStateProviderInternal.isInAmbientMode = false
        updateDisplay()
    }

    override fun onResume() {
        super.onResume()
        app.onResume()
        app.messageExchangeHelper.addMessageListenerWeakly(this)
        val navigator = app.musicLibraryNavigator
        navigator.onFileSelected += mFileSelectedEventHandler
        navigator.onAlbumSelected += mItemSelectedEventHandler
        navigator.onFolderSelected += mItemSelectedEventHandler
        navigator.onCategorySelected += mItemSelectedEventHandler
    }

    override fun onPause() {
        app.onPause()
        app.messageExchangeHelper.removeMessageListener(this)
        val navigator = app.musicLibraryNavigator
        navigator.onFileSelected -= mFileSelectedEventHandler
        navigator.onAlbumSelected -= mItemSelectedEventHandler
        navigator.onFolderSelected -= mItemSelectedEventHandler
        navigator.onCategorySelected -= mItemSelectedEventHandler
        super.onPause()
    }

    override fun onMessageReceived(messageEvent: MessageEvent) {
        if (AlbumArtChangedEvent.PATH == messageEvent.path) {
            val bytes = messageEvent.data
            if (bytes === null) {
                mAlbumArtBitmap = null
            } else {
                mAlbumArtBitmap = BitmapFactory.decodeByteArray(bytes, 0, bytes.size)
            }
            updateDisplay()
        } else if (TrackChangedEvent.PATH == messageEvent.path && mShouldNavigateToPlayerView) {
            mShouldNavigateToPlayerView = false
            mViewPager!!.setCurrentItem(0, true)
        }
    }

    private fun updateDisplay() {
        mAlbumArt!!.setImageBitmap(if (isAmbient) null else mAlbumArtBitmap)
    }

}

internal val Context.ambientModeStateProviderInternal: AmbientModeStateProviderImpl
    get() =  ambientModeStateProvider as AmbientModeStateProviderImpl
